# House Prices - Advanced Regression Techniques
This is my second submission in Kaggle so if you find any mistakes, corrections, or suggestions, feel free to mention them.

## Initial Setup

### Import Statements

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### Import the Dataset

In [2]:
train_data = pd.read_csv("./datasets/train.csv")
test_data = pd.read_csv("./datasets/test.csv")

### Stylistic

In [3]:
%matplotlib inline
pd.set_option("display.max_rows", 500)
pd.set_option("display.max_columns", 500)
pd.set_option("display.width", 500)
pd.set_option("precision", 2)
plt.style.use("dark_background")
np.set_printoptions(precision=2)

### Files

In [4]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('./datasets/'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

./datasets/test.csv
./datasets/data_description.txt
./datasets/train.csv
./datasets/sample_submission.csv


### Exploring the Dataset

#### Description of Each Column

In [5]:
train_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2003          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace       196.0        Gd        TA      PConc       Gd       TA           No          GLQ         706          Unf           0        150          856    GasA        Ex          Y      SBrkr       856       854             0   
1   2          20       RL         80.0     9600   Pave   NaN      Reg         Lvl    AllPub       FR2       Gtl      Veenker      Feedr       Norm     1Fam     1Story            6            8       1976          1976     Gable  CompShg     MetalSd     MetalSd       None         0.0        TA        TA     CBlock       Gd       TA           Gd          ALQ         978          Unf           0        284         1262    GasA        Ex          Y      SBrkr      1262         0             0   
2   3          60       RL         68.0    11250   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl      CollgCr       Norm       Norm     1Fam     2Story            7            5       2001          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace       162.0        Gd        TA      PConc       Gd       TA           Mn          GLQ         486          Unf           0        434          920    GasA        Ex          Y      SBrkr       920       866             0   
3   4          70       RL         60.0     9550   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl      Crawfor       Norm       Norm     1Fam     2Story            7            5       1915          1970     Gable  CompShg     Wd Sdng     Wd Shng       None         0.0        TA        TA     BrkTil       TA       Gd           No          ALQ         216          Unf           0        540          756    GasA        Gd          Y      SBrkr       961       756             0   
4   5          60       RL         84.0    14260   Pave   NaN      IR1         Lvl    AllPub       FR2       Gtl      NoRidge       Norm       Norm     1Fam     2Story            8            5       2000          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace       350.0        Gd        TA      PConc       Gd       TA           Av          GLQ         655          Unf           0        490         1145    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  SalePrice  
0       1710             1             0         2         1             3             1          Gd             8        Typ           0         NaN     Attchd       2003.0          RFn           2         548         TA         TA          Y           0           61              0          0            0         0    NaN   NaN         NaN        0       2    2008       WD        Normal     208500  
1       1262             0             1         2         0             3             1          TA             6        Typ           1          TA     Attchd       1976.0          RFn           2        

In [6]:
test_data.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape LandContour Utilities LotConfig LandSlope Neighborhood Condition1 Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg         Lvl    AllPub    Inside       Gtl        NAmes      Feedr       Norm     1Fam     1Story            5            6       1961          1961     Gable  CompShg     VinylSd     VinylSd       None         0.0        TA        TA     CBlock       TA       TA           No          Rec       468.0          LwQ       144.0      270.0        882.0    GasA        TA          Y      SBrkr       896         0             0   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1         Lvl    AllPub    Corner       Gtl        NAmes       Norm       Norm     1Fam     1Story            6            6       1958          1958       Hip  CompShg     Wd Sdng     Wd Sdng    BrkFace       108.0        TA        TA     CBlock       TA       TA           No          ALQ       923.0          Unf         0.0      406.0       1329.0    GasA        TA          Y      SBrkr      1329         0             0   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl      Gilbert       Norm       Norm     1Fam     2Story            5            5       1997          1998     Gable  CompShg     VinylSd     VinylSd       None         0.0        TA        TA      PConc       Gd       TA           No          GLQ       791.0          Unf         0.0      137.0        928.0    GasA        Gd          Y      SBrkr       928       701             0   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1         Lvl    AllPub    Inside       Gtl      Gilbert       Norm       Norm     1Fam     2Story            6            6       1998          1998     Gable  CompShg     VinylSd     VinylSd    BrkFace        20.0        TA        TA      PConc       TA       TA           No          GLQ       602.0          Unf         0.0      324.0        926.0    GasA        Ex          Y      SBrkr       926       678             0   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1         HLS    AllPub    Inside       Gtl      StoneBr       Norm       Norm   TwnhsE     1Story            8            5       1992          1992     Gable  CompShg     HdBoard     HdBoard       None         0.0        Gd        TA      PConc       Gd       TA           No          ALQ       263.0          Unf         0.0     1017.0       1280.0    GasA        Ex          Y      SBrkr      1280         0             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu GarageType  GarageYrBlt GarageFinish  GarageCars  GarageArea GarageQual GarageCond PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea PoolQC  Fence MiscFeature  MiscVal  MoSold  YrSold SaleType SaleCondition  
0        896           0.0           0.0         1         0             2             1          TA             5        Typ           0         NaN     Attchd       1961.0          Unf         1.0       730.0         TA         TA          Y         140            0              0          0          120         0    NaN  MnPrv         NaN        0       6    2010       WD        Normal  
1       1329           0.0           0.0         1         1             3             1          Gd             6        Typ           0         NaN     Attchd       1958.0          Unf         1.0       312.0      

In [7]:
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

Some insights for this data are:
1. We have 1460 entries.
2. We have 81 Columns.
3. The Indexing for this column are from 0 to 1459.
4. Alley, FireplaceQu, PoolQC, Fence, MiscFeature are some column with sparse data (Most data are Zero).
5. Most of our data are string, which needed to be encoded.

## Data Cleaning

### Inspecting Null Values

Removing null values is an essential step to train our model to output accurate results without error. For that, let's inspect each columns and their sum of null values.

In [8]:
# Train Data
train_nulls = train_data.isnull().sum()
train_percent = (train_data.isnull().sum()/train_data.isnull().count())
train_missing_vals = pd.concat([train_nulls, train_percent], axis=1, keys=["Train_Total", "Train_Pecentage"]).sort_values(by="Train_Total", ascending=False)

# Test Data
test_nulls = test_data.isnull().sum()
test_percent = (test_data.isnull().sum()/test_data.isnull().count())
test_missing_vals = pd.concat([test_nulls, test_percent], axis=1, keys=["Test_Total", "Test_Pecentage"]).sort_values(by="Test_Total", ascending=False)

total_null = pd.concat([train_missing_vals, test_missing_vals], axis=1, keys=["Train_Total", "Train_Pecentage", "Test_Total", "Test_Pecentage"])
total_null.columns = total_null.columns.droplevel()
total_null.head(45)

Train_Total  Train_Pecentage  Test_Total  Test_Pecentage
PoolQC                1453         9.95e-01      1456.0        9.98e-01
MiscFeature           1406         9.63e-01      1408.0        9.65e-01
Alley                 1369         9.38e-01      1352.0        9.27e-01
Fence                 1179         8.08e-01      1169.0        8.01e-01
FireplaceQu            690         4.73e-01       730.0        5.00e-01
LotFrontage            259         1.77e-01       227.0        1.56e-01
GarageYrBlt             81         5.55e-02        78.0        5.35e-02
GarageCond              81         5.55e-02        78.0        5.35e-02
GarageType              81         5.55e-02        76.0        5.21e-02
GarageFinish            81         5.55e-02        78.0        5.35e-02
GarageQual              81         5.55e-02        78.0        5.35e-02
BsmtFinType2            38         2.60e-02        42.0        2.88e-02
BsmtExposure            38         2.60e-02        44.0        3.02e-02
BsmtQual                37         2.53e-02        44.0        3.02e-02
BsmtCond                37         2.53e-02        45.0        3.08e-02
BsmtFinType1            37         2.53e-02        42.0        2.88e-02
MasVnrArea               8         5.48e-03        15.0        1.03e-02
MasVnrType               8         5.48e-03        16.0        1.10e-02
Electrical               1         6.85e-04         0.0        0.00e+00
Id                       0         0.00e+00         0.0        0.00e+00
Functional               0         0.00e+00         2.0        1.37e-03
Fireplaces               0         0.00e+00         0.0        0.00e+00
KitchenQual              0         0.00e+00         1.0        6.85e-04
KitchenAbvGr             0         0.00e+00         0.0        0.00e+00
BedroomAbvGr             0         0.00e+00         0.0        0.00e+00
HalfBath                 0         0.00e+00         0.0        0.00e+00
FullBath                 0         0.00e+00         0.0        0.00e+00
BsmtHalfBath             0         0.00e+00         2.0        1.37e-03
TotRmsAbvGrd             0         0.00e+00         0.0        0.00e+00
GarageCars               0         0.00e+00         1.0        6.85e-04
GrLivArea                0         0.00e+00         0.0        0.00e+00
GarageArea               0         0.00e+00         1.0        6.85e-04
PavedDrive               0         0.00e+00         0.0        0.00e+00
WoodDeckSF               0         0.00e+00         0.0        0.00e+00
OpenPorchSF              0         0.00e+00         0.0        0.00e+00
EnclosedPorch            0         0.00e+00         0.0        0.00e+00
3SsnPorch                0         0.00e+00         0.0        0.00e+00
ScreenPorch              0         0.00e+00         0.0        0.00e+00
PoolArea                 0         0.00e+00         0.0        0.00e+00
MiscVal                  0         0.00e+00         0.0        0.00e+00
MoSold                   0         0.00e+00         0.0        0.00e+00
YrSold                   0         0.00e+00         0.0        0.00e+00
SaleType                 0         0.00e+00         1.0        6.85e-04
SaleCondition            0         0.00e+00         0.0        0.00e+00
BsmtFullBath             0         0.00e+00         2.0        1.37e-03

#### Category Values

In [9]:
# Train Data
null_cat = train_missing_vals[train_missing_vals["Train_Total"] >=  1].index

for cols in null_cat[:5]:
    print(train_data[cols].value_counts().head(), "\n")

Gd    3
Fa    2
Ex    2
Name: PoolQC, dtype: int64 

Shed    49
Othr     2
Gar2     2
TenC     1
Name: MiscFeature, dtype: int64 

Grvl    50
Pave    41
Name: Alley, dtype: int64 

MnPrv    157
GdPrv     59
GdWo      54
MnWw      11
Name: Fence, dtype: int64 

Gd    380
TA    313
Fa     33
Ex     24
Po     20
Name: FireplaceQu, dtype: int64 



In [10]:
# Test Data
null_cat = test_missing_vals[test_missing_vals["Test_Total"] >=  1].index

for cols in null_cat[:5]:
    print(test_data[cols].value_counts(), "\n")

Ex    2
Gd    1
Name: PoolQC, dtype: int64 

Shed    46
Gar2     3
Othr     2
Name: MiscFeature, dtype: int64 

Grvl    70
Pave    37
Name: Alley, dtype: int64 

MnPrv    172
GdPrv     59
GdWo      58
MnWw       1
Name: Fence, dtype: int64 

Gd    364
TA    279
Fa     41
Po     26
Ex     19
Name: FireplaceQu, dtype: int64 



It seems that some of the values for every `object` column  doesn't have "NA" as opposed to what have been said in the data description. Let's fix this by filling the null values with "None"

In [11]:
train_data.loc[:, train_data.dtypes == "object"] = train_data.loc[:, train_data.dtypes == "object"].fillna("NA")

In [12]:
test_data.loc[:, test_data.dtypes == "object"] = test_data.loc[:, test_data.dtypes == "object"].fillna("NA")

Now we can confirm that all the missing values had been filled with "None" for the Category Columns.

In [13]:
train_data.loc[:, train_data.dtypes == "object"].isnull().sum().max()

0

In [14]:
test_data.loc[:, test_data.dtypes == "object"].isnull().sum().max()

0

#### Numerical Values

In [15]:
train_data.loc[:, train_data.dtypes != "object"].isnull().sum().sort_values(ascending=False).head()

LotFrontage    259
GarageYrBlt     81
MasVnrArea       8
Id               0
OpenPorchSF      0
dtype: int64

It seems only 3 columns does have a null values in numerical column. Let's fill these null values based on their data description.

In [16]:
train_data["LotFrontage"].fillna(train_data["LotFrontage"].mean(), inplace=True)
train_data["GarageYrBlt"].fillna(train_data["GarageYrBlt"].median(), inplace=True)
train_data["MasVnrArea"].fillna(train_data["MasVnrArea"].mean(), inplace=True)

Here we filled `GarageYrBlt` with median as the values for these columns are `Year` and the rest are measurements so we get their mean and filled those with mean values.

Now, let's check the maximum missing values for train_data.

In [17]:
train_data.loc[:, train_data.dtypes != "object"].isnull().sum().max()

0

In [18]:
test_data.loc[:, test_data.dtypes != "object"].isnull().sum().sort_values(ascending=False).head(15)

LotFrontage      227
GarageYrBlt       78
MasVnrArea        15
BsmtHalfBath       2
BsmtFullBath       2
BsmtFinSF2         1
GarageCars         1
GarageArea         1
TotalBsmtSF        1
BsmtUnfSF          1
BsmtFinSF1         1
KitchenAbvGr       0
3SsnPorch          0
EnclosedPorch      0
OpenPorchSF        0
dtype: int64

We can observer  here that the three columns we have filled in train_data (LotFrontage, GarageYrBlt, MasVnrArea)  has missing values here as well, let's copy and paste the code above and deal with this just like before. For the other columns, let's check on what values it holds to fill these values appropriately.

In [19]:
test_data["LotFrontage"].fillna(test_data["LotFrontage"].mean(), inplace=True)
test_data["GarageYrBlt"].fillna(test_data["GarageYrBlt"].median(), inplace=True)
test_data["MasVnrArea"].fillna(test_data["MasVnrArea"].mean(), inplace=True)

In [20]:
cols = ["BsmtHalfBath", "BsmtFullBath", "BsmtFinSF2", "GarageCars", "GarageArea", "TotalBsmtSF", "BsmtUnfSF", "BsmtFinSF1"]
for col in cols:
    print(test_data.loc[:10, col].value_counts(dropna=False), "\n")

0.0    11
Name: BsmtHalfBath, dtype: int64 

0.0    7
1.0    4
Name: BsmtFullBath, dtype: int64 

0.0      9
144.0    1
78.0     1
Name: BsmtFinSF2, dtype: int64 

2.0    9
1.0    2
Name: GarageCars, dtype: int64 

506.0    2
482.0    1
312.0    1
440.0    1
730.0    1
393.0    1
525.0    1
511.0    1
470.0    1
420.0    1
Name: GarageArea, dtype: int64 

882.0     2
1300.0    1
1280.0    1
926.0     1
789.0     1
928.0     1
763.0     1
1405.0    1
1329.0    1
1168.0    1
Name: TotalBsmtSF, dtype: int64 

233.0     1
763.0     1
324.0     1
354.0     1
789.0     1
270.0     1
1017.0    1
406.0     1
137.0     1
663.0     1
0.0       1
Name: BsmtUnfSF, dtype: int64 

0.0       2
791.0     1
935.0     1
602.0     1
923.0     1
637.0     1
804.0     1
263.0     1
468.0     1
1051.0    1
Name: BsmtFinSF1, dtype: int64 



Since some of the columns have values `0.0` (0.0) while some are measurements (Mean) and others are numerical order (Mode), we will handle each column differently

In [21]:
mean_ = ["BsmtFinSF2", "GarageArea", "GarageArea"]
mode_ = ["BsmtHalfBath", "BsmtFullBath"]
zero_ = ["GarageCars", "BsmtUnfSF", "TotalBsmtSF", "BsmtFinSF1"]

for col1 in mean_:
    test_data[col1] = test_data[col1].fillna(test_data[col1].mean())

for col2 in mode_:
    test_data[col2] = test_data[col2].fillna(test_data[col2].mode()[0])
    
for col3 in zero_:
    test_data[col3] = test_data[col3].fillna(0.0)

Same procedure for test_data. We checked the columns that have null values and fill those with appropriate values; either mean or mode or a 0.0.

In [22]:
test_data.isnull().sum().max()

0

Now we are sure that everything in train_data and test_data have no null values.

### Removing Outliers

#### Get the Numeric and Category Columns

In [23]:
numeric_col_train = [col for col in train_data.columns if train_data[col].dtype == "int64" or train_data[col].dtype == "float64"][1:]
category_cols_train = [col for col in train_data.columns if train_data[col].dtype == "object"]

numeric_col_test = [col for col in test_data.columns if test_data[col].dtype == "int64" or train_data[col].dtype == "float64"][1:]
category_cols_test = [col for col in test_data.columns if test_data[col].dtype == "object"]

#### Train Data

In [24]:
before_removal = train_data.shape[0]

for feature in numeric_col_train:
    min_ = train_data[feature].mean() - (3 * train_data[feature].std())
    max_ = train_data[feature].mean() + (3 * train_data[feature].std())
    
    train_data = train_data[train_data[feature] >= min_]
    train_data = train_data[train_data[feature] <= max_]
    
print("Number of Observation Removed: {}".format(before_removal - train_data.shape[0]))

Number of Observation Removed: 492


## Data Preprocessing

### LabelEncoder()

#### Train Data

In [25]:
from sklearn.preprocessing import LabelEncoder

enc = LabelEncoder()
for feature in category_cols_train:
    train_data[feature] = enc.fit_transform(train_data[feature].values)

In [26]:
train_data.head()

Id  MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  LandContour  Utilities  LotConfig  LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  Exterior2nd  MasVnrType  MasVnrArea  ExterQual  ExterCond  Foundation  BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  BsmtFinSF1  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  Heating  HeatingQC  CentralAir  Electrical  \
0    1          60         3         65.0     8450       1      1         3            3          0          4          0             5           2           1         0           4            7            5       2003          2003          1         0            9           11           1       196.0          2          2           2         2         3             4             2         706             6           0        150          856        1          0           1           4   
2    3          60         3         68.0    11250       1      1         0            3          0          4          0             5           2           1         0           4            7            5       2001          2002          1         0            9           11           1       162.0          2          2           2         2         3             2             2         486             6           0        434          920        1          0           1           4   
4    5          60         3         84.0    14260       1      1         0            3          0          2          0            15           2           1         0           4            8            5       2000          2000          1         0            9           11           1       350.0          2          2           2         2         3             0             2         655             6           0        490         1145        1          0           1           4   
6    7          20         3         75.0    10084       1      1         3            3          0          4          0            21           2           1         0           2            8            5       2004          2005          1         0            9           11           4       186.0          2          2           2         0         3             0             2        1369             6           0        317         1686        1          0           1           4   
10  11          20         3         70.0    11200       1      1         3            3          0          4          0            19           2           1         0           2            5            5       1965          1965          3         0            4            4           3         0.0          3          2           1         4         3             4             5         906             6           0        134         1040        1          0           1           4   

    1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  Functional  Fireplaces  FireplaceQu  GarageType  GarageYrBlt  GarageFinish  GarageCars  GarageArea  GarageQual  GarageCond  PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  PoolQC  Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  SalePrice  
0        856       854             0       1710             1             0         2         1             3             1            2             8           5           0            3           0       2003.0             2           2         548           5           5           2           0           61              0          0            0         0       0      4            0        0       2    2008         8              4     208500  
2        920       866             0       1786             1             0         2         1             3             1      

#### Test Data

In [27]:
from sklearn.preprocessing import LabelEncoder

enc = LabelEncoder()
for feature in category_cols_test:
    test_data[feature] = enc.fit_transform(test_data[feature].values)

In [28]:
test_data.head()

Id  MSSubClass  MSZoning  LotFrontage  LotArea  Street  Alley  LotShape  LandContour  Utilities  LotConfig  LandSlope  Neighborhood  Condition1  Condition2  BldgType  HouseStyle  OverallQual  OverallCond  YearBuilt  YearRemodAdd  RoofStyle  RoofMatl  Exterior1st  Exterior2nd  MasVnrType  MasVnrArea  ExterQual  ExterCond  Foundation  BsmtQual  BsmtCond  BsmtExposure  BsmtFinType1  BsmtFinSF1  BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  Heating  HeatingQC  CentralAir  Electrical  \
0  1461          20         3         80.0    11622       1      1         3            3          0          4          0            12           1           2         0           2            5            6       1961          1961          1         0           11           13           3         0.0          3          4           1         4         4             4             5       468.0             3       144.0      270.0        882.0        0          4           1           3   
1  1462          20         4         81.0    14267       1      1         0            3          0          0          0            12           2           2         0           2            6            6       1958          1958          3         0           12           14           1       108.0          3          4           1         4         4             4             0       923.0             6         0.0      406.0       1329.0        0          4           1           3   
2  1463          60         4         74.0    13830       1      1         0            3          0          4          0             8           2           2         0           4            5            5       1997          1998          1         0           11           13           3         0.0          3          4           2         2         4             4             2       791.0             6         0.0      137.0        928.0        0          2           1           3   
3  1464          60         4         78.0     9978       1      1         0            3          0          4          0             8           2           2         0           4            6            6       1998          1998          1         0           11           13           1        20.0          3          4           2         4         4             4             2       602.0             6         0.0      324.0        926.0        0          0           1           3   
4  1465         120         4         43.0     5005       1      1         0            1          0          4          0            22           2           2         4           2            8            5       1992          1992          1         0            6            6           3         0.0          2          4           2         2         4             4             0       263.0             6         0.0     1017.0       1280.0        0          0           1           3   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  KitchenQual  TotRmsAbvGrd  Functional  Fireplaces  FireplaceQu  GarageType  GarageYrBlt  GarageFinish  GarageCars  GarageArea  GarageQual  GarageCond  PavedDrive  WoodDeckSF  OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  PoolQC  Fence  MiscFeature  MiscVal  MoSold  YrSold  SaleType  SaleCondition  
0       896         0             0        896           0.0           0.0         1         0             2             1            4             5           7           0            3           1       1961.0             3         1.0       730.0           4           5           2         140            0              0          0          120         0       2      2            1        0       6    2010         9              4  
1      1329         0             0       1329           0.0           0.0         1         1             3             1            2             

## Machine Learning Model

### Splitting the Dataset

In [29]:
X = train_data.drop(["Id", "SalePrice"], axis=1).values
y = train_data["SalePrice"].values

Here we also removed `Id` from `X` since `Id` is only an identifier and doesn't have anything to do with the dependent variable.

In [30]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y)

### Feature Scaling

In [31]:
from sklearn.preprocessing import StandardScaler
enc = StandardScaler()
# Train Data
X_train = enc.fit_transform(X_train)
# Test Data
X_test = enc.transform(X_test)

### Fitting the Training Data to Model

In [32]:
from sklearn.model_selection import cross_val_score # Scoring Metric

#### Random Forest

In [33]:
from sklearn.ensemble import RandomForestRegressor

forest = RandomForestRegressor(max_depth=6, n_estimators=200, n_jobs=-1).fit(X_train, y_train)
acc_forest = forest.score(X_train, y_train)*100
acc_test_forest = forest.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_forest))
print("Accuracy on Test Set: {:.2f}".format(acc_test_forest))

# k-Fold Metric
kfold_forest = cross_val_score(estimator = forest, X = X, y = y, cv=10)
kfold_forest_mean = kfold_forest.mean()*100
print("k-Fold: {:.2f}%".format(kfold_forest_mean))
print("Standard Deviation: {:.2f}%".format(kfold_forest.std()*100))

Accuracy on Training Set: 95.24
Accuracy on Test Set: 88.76
k-Fold: 87.99%
Standard Deviation: 2.21%


#### Gradient Boosting Regressor

In [34]:
from sklearn.ensemble import GradientBoostingRegressor

gradient = GradientBoostingRegressor(max_depth=1, learning_rate=0.29, max_features=10, random_state=0).fit(X_train, y_train)
acc_gradient = gradient.score(X_train, y_train)*100
acc_test_gradient = gradient.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_gradient))
print("Accuracy on Test Set: {:.2f}".format(acc_test_gradient))

# k-Fold Metric
kfold_gradient = cross_val_score(estimator = gradient, X = X, y = y, cv=10)
kfold_gradient_mean = kfold_gradient.mean()*100
print("k-Fold: {:.2f}%".format(kfold_gradient_mean))
print("Standard Deviation: {:.2f}%".format(kfold_gradient.std()*100))

Accuracy on Training Set: 92.49
Accuracy on Test Set: 89.62
k-Fold: 88.84%
Standard Deviation: 1.98%


#### Suport Vector Regressor

In [35]:
from sklearn.svm import SVR

svr = SVR(kernel = "rbf").fit(X_train, y_train)
acc_svr = svr.score(X_train, y_train)*100
acc_test_svr = svr.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_svr))
print("Accuracy on Test Set: {:.2f}".format(acc_test_svr))

# k-Fold Metric
kfold_svr = cross_val_score(estimator = svr, X = X, y = y, cv=10)
kfold_svr_mean = kfold_svr.mean()*100
print("k-Fold: {:.2f}%".format(kfold_svr_mean))
print("Standard Deviation: {:.2f}%".format(kfold_svr.std()*100))

Accuracy on Training Set: -3.81
Accuracy on Test Set: -7.59
k-Fold: -3.94%
Standard Deviation: 3.04%


#### KNN Regressor

In [36]:
from sklearn.neighbors import KNeighborsRegressor

knn = KNeighborsRegressor().fit(X_train, y_train)
acc_knn = knn.score(X_train, y_train)*100
acc_test_knn = knn.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_knn))
print("Accuracy on Test Set: {:.2f}".format(acc_test_knn))

# k-Fold Metric
kfold_knn = cross_val_score(estimator = knn, X = X, y = y, cv=10)
kfold_knn_mean = kfold_knn.mean()*100
print("k-Fold: {:.2f}%".format(kfold_knn_mean))
print("Standard Deviation: {:.2f}%".format(kfold_knn.std()*100))

Accuracy on Training Set: 88.55
Accuracy on Test Set: 84.91
k-Fold: 71.81%
Standard Deviation: 5.39%


#### Linear Regressor

In [37]:
from sklearn.linear_model import LinearRegression

linearReg = LinearRegression().fit(X_train, y_train)
acc_linearReg = linearReg.score(X_train, y_train)*100
acc_test_linearReg = linearReg.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_linearReg))
print("Accuracy on Test Set: {:.2f}".format(acc_test_linearReg))

# k-Fold Metric
kfold_linearReg = cross_val_score(estimator = linearReg, X = X, y = y, cv=10)
kfold_linearReg_mean = kfold_linearReg.mean()*100
print("k-Fold: {:.2f}%".format(kfold_linearReg_mean))
print("Standard Deviation: {:.2f}%".format(kfold_linearReg.std()*100))

Accuracy on Training Set: 91.79
Accuracy on Test Set: 91.01
k-Fold: 89.62%
Standard Deviation: 1.69%


#### Linear Support Vector Regressor

In [38]:
from sklearn.svm import LinearSVR

linearSVR = LinearSVR().fit(X_train, y_train)
acc_linearSVR = linearSVR.score(X_train, y_train)*100
acc_test_linearSVR = linearSVR.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_linearSVR))
print("Accuracy on Test Set: {:.2f}".format(acc_test_linearSVR))

# k-Fold Metric
kfold_linearSVR = cross_val_score(estimator = linearSVR, X = X, y = y, cv=10)
kfold_linearSVR_mean = kfold_linearSVR.mean()*100
print("k-Fold: {:.2f}%".format(kfold_linearSVR_mean))
print("Standard Deviation: {:.2f}%".format(kfold_linearSVR.std()*100))

Accuracy on Training Set: -788.18
Accuracy on Test Set: -845.80


/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/home/prince/Github/Kaggle Compe

k-Fold: 64.37%
Standard Deviation: 27.24%


/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "
/home/prince/Github/Kaggle Competitions/venv/lib/python3.8/site-packages/sklearn/svm/_base.py:985: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn("Liblinear failed to converge, increase "


#### Decision Tree

In [39]:
from sklearn.tree import DecisionTreeRegressor

tree = DecisionTreeRegressor(max_depth=5, random_state=0).fit(X_train, y_train)
acc_tree = tree.score(X_train, y_train)*100
acc_test_tree = tree.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_tree))
print("Accuracy on Test Set: {:.2f}".format(acc_test_tree))

# k-Fold Metric
kfold_tree = cross_val_score(estimator = tree, X = X, y = y, cv=10)
kfold_tree_mean = kfold_tree.mean()*100
print("k-Fold: {:.2f}%".format(kfold_tree_mean))
print("Standard Deviation: {:.2f}%".format(kfold_tree.std()*100))

Accuracy on Training Set: 88.45
Accuracy on Test Set: 74.84
k-Fold: 78.57%
Standard Deviation: 3.83%


#### Ridge Regressor

In [40]:
from sklearn.linear_model import Ridge

ridge = Ridge(alpha=1).fit(X_train, y_train)
acc_ridge = tree.score(X_train, y_train)*100
acc_test_ridge = tree.score(X_test, y_test)*100

print("Accuracy on Training Set: {:.2f}".format(acc_ridge))
print("Accuracy on Test Set: {:.2f}".format(acc_test_ridge))

# k-Fold Metric
kfold_ridge = cross_val_score(estimator = ridge, X = X, y = y, cv=10)
kfold_ridge_mean = kfold_ridge.mean()*100
print("k-Fold: {:.2f}%".format(kfold_ridge_mean))
print("Standard Deviation: {:.2f}%".format(kfold_ridge.std()*100))

Accuracy on Training Set: 88.45
Accuracy on Test Set: 74.84
k-Fold: 89.88%
Standard Deviation: 1.47%


#### Polynomial Linear Regressor

In [41]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression

poly_reg = PolynomialFeatures(interaction_only=True)
X_poly = poly_reg.fit_transform(X_train)
reg = LinearRegression()
_ = reg.fit(X_poly, y_train)

acc_poly = reg.score(X_poly, y_train)*100
# acc_test_poly = reg.score(X_train, poly_reg.transform(X_test))*100

print("Accuracy on Training Set: {:.2f}".format(acc_poly))
# print("Accuracy on Training Set: {:.2f}".format(acc_test_poly))

# k-Fold Metric
kfold_poly = cross_val_score(estimator = reg, X = X, y = y, cv=10)
kfold_poly_mean = kfold_poly.mean()*100
print("k-Fold: {:.2f}%".format(kfold_poly_mean))
print("Standard Deviation: {:.2f}%".format(kfold_poly.std()*100))

Accuracy on Training Set: 100.00
k-Fold: 89.62%
Standard Deviation: 1.69%


#### XGBoost Regressor

In [42]:
from xgboost import XGBRegressor
xgb = XGBRegressor(min_child_weight = 3, 
                   max_depth = 10, 
                   learning_rate = 0.05, 
                   gamma = 0.2,
                   colsample_bytree = 0.4).fit(X_train, y_train)

# R2 Score Metric
acc_xgb = xgb.score(X_train, y_train)*100
acc_test_xgb = xgb.score(X_test, y_test)*100
print("Accuracy on Training Set: {:.2f}".format(acc_xgb))
print("Accuracy on Test Set: {:.2f}".format(acc_test_xgb))

# k-Fold Metric
kfold_xgb = cross_val_score(estimator = xgb, X = X, y = y, cv=10)
kfold_xgb_mean = round(kfold_xgb.mean()*100, 2)
print("k-Fold: {}%".format(kfold_xgb_mean))
print("Standard Deviation: {:.2f}%".format(kfold_xgb.std()*100))

Accuracy on Training Set: 99.52
Accuracy on Test Set: 91.21
k-Fold: 90.32%
Standard Deviation: 1.54%


### Hyperparameter Tuning

In [43]:
# from sklearn.model_selection import GridSearchCV
# from sklearn.metrics import make_scorer, mean_squared_error
# # Hyper Parameter Optimization

# params={
#  "learning_rate"    : [0.05, 0.10, 0.15, 0.20, 0.25, 0.30 ],
#  "max_depth"        : [ 3, 4, 5, 6, 8, 10, 12, 15],
#  "min_child_weight" : [ 1, 3, 5, 7 ],
#  "gamma"            : [ 0.0, 0.1, 0.2 , 0.3, 0.4 ],
#  "colsample_bytree" : [ 0.3, 0.4, 0.5 , 0.7 ]
# }

# MSE = make_scorer(mean_squared_error, greater_is_better=False)
# grid_search = GridSearchCV(estimator = xgb,
#                           param_grid = params,
#                           cv = 4,
#                           scoring = MSE,
#                           n_jobs = -1,
#                           verbose = 2)
# grid_search.fit(X, y)
# best_accuracy = grid_search.best_score_
# best_params = grid_search.best_params_
# print("Best Accuracy: {:.2f}".format(best_accuracy*100))
# print("Best Parameters:", best_params)

### Score Metrics

In [44]:
models = pd.DataFrame({
    
    "Model": ["KNN", "Linear Regression", "Random Forest", "Linear SVC", "Decision Tree", "Ridge Regression", "Gradient Boosting Regressor", "XGBoost Regressor"],
    "Train Score": [acc_knn, acc_linearReg, acc_forest, acc_linearSVR, acc_tree, acc_ridge, acc_gradient, acc_xgb],
    "Test Score": [acc_test_knn, acc_test_linearReg, acc_test_forest, acc_test_linearSVR, acc_test_tree, acc_test_ridge, acc_test_gradient, acc_test_xgb],
    "K-Fold": [kfold_knn_mean, kfold_linearReg_mean, kfold_forest_mean, kfold_linearSVR_mean, kfold_tree_mean, kfold_ridge_mean, kfold_gradient_mean, kfold_xgb_mean]
    
                     })
sorted_models = models.sort_values(by=["Train Score", "Test Score"], ascending=False)

In [45]:
sorted_models

Model  Train Score  Test Score  K-Fold
7            XGBoost Regressor        99.52       91.21   90.32
2                Random Forest        95.24       88.76   87.99
6  Gradient Boosting Regressor        92.49       89.62   88.84
1            Linear Regression        91.79       91.01   89.62
0                          KNN        88.55       84.91   71.81
4                Decision Tree        88.45       74.84   78.57
5             Ridge Regression        88.45       74.84   89.88
3                   Linear SVC      -788.18     -845.80   64.37

### Submission Predictions

In [46]:
predictions = xgb.predict(enc.transform(test_data.drop("Id", axis=1)))

In [47]:
predictions

array([122895.44, 145486.77, 176527.77, ..., 149523.61, 112957.43,
       228906.5 ], dtype=float32)

In [48]:
output = pd.DataFrame({'Id': test_data["Id"], 'SalePrice': predictions})
output.to_csv('my_submission.csv', index=False)
print("Your submission was successfully saved!")

Your submission was successfully saved!
